File: src/test/java/com/flexpoker/table/command/aggregate/testhelpers/TableTestUtils.java
package com.flexpoker.table.command.aggregate.testhelpers;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.UUID;

import com.flexpoker.model.HandRanking;
import com.flexpoker.model.card.Card;
import com.flexpoker.model.card.CardRank;
import com.flexpoker.model.card.PocketCards;
import com.flexpoker.table.command.aggregate.DefaultTableFactory;
import com.flexpoker.table.command.aggregate.HandEvaluation;
import com.flexpoker.table.command.aggregate.Table;
import com.flexpoker.table.command.commands.CreateTableCommand;
import com.flexpoker.table.command.events.HandDealtEvent;
import com.flexpoker.table.command.events.TableCreatedEvent;
import com.flexpoker.test.util.datageneration.CardGenerator;
import com.flexpoker.test.util.datageneration.DeckGenerator;

public class TableTestUtils {

    public static Table createBasicTable(UUID tableId, UUID... playerIdsArray) {
        var playerIds = new HashSet<>(Arrays.asList(playerIdsArray));
        var command = new CreateTableCommand(tableId, UUID.randomUUID(), playerIds, 6);
        return new DefaultTableFactory().createNew(command);
    }

    public static Table createBasicTableAndStartHand(UUID tableId, UUID... playerIdsArray) {
        var playerIds = new HashSet<>(Arrays.asList(playerIdsArray));

        var smallBlind = 10;
        var bigBlind = 20;
        var shuffledDeckOfCards = new ArrayList<Card>();
        var cardsUsedInHand = DeckGenerator.createDeck();

        var handEvaluation1 = new HandEvaluation();
        handEvaluation1.setPlayerId(playerIdsArray[0]);
        handEvaluation1.setHandRanking(HandRanking.FLUSH);
        handEvaluation1.setPrimaryCardRank(CardRank.KING);

        var handEvaluation2 = new HandEvaluation();
        handEvaluation2.setPlayerId(playerIdsArray[1]);
        handEvaluation2.setHandRanking(HandRanking.STRAIGHT);
        handEvaluation2.setPrimaryCardRank(CardRank.KING);

        var handEvaluations = new HashMap<PocketCards, HandEvaluation>();
        handEvaluations.put(CardGenerator.createPocketCards1(), handEvaluation1);
        handEvaluations.put(CardGenerator.createPocketCards2(), handEvaluation2);

        var command = new CreateTableCommand(tableId, UUID.randomUUID(), playerIds, 6);
        var table = new DefaultTableFactory().createNew(command);
        table.startNewHandForNewGame(smallBlind, bigBlind, shuffledDeckOfCards, cardsUsedInHand, handEvaluations);
        return table;
    }

    public static UUID fetchIdForButton(TableCreatedEvent tableCreatedEvent,
            HandDealtEvent handDealtEvent) {
        var seatPositionToPlayerIdMap = tableCreatedEvent.getSeatPositionToPlayerMap();
        return seatPositionToPlayerIdMap.get(handDealtEvent.getButtonOnPosition());
    }

    public static UUID fetchIdForSmallBlind(TableCreatedEvent tableCreatedEvent,
            HandDealtEvent handDealtEvent) {
        var seatPositionToPlayerIdMap = tableCreatedEvent.getSeatPositionToPlayerMap();
        return seatPositionToPlayerIdMap.get(handDealtEvent.getSmallBlindPosition());
    }

    public static UUID fetchIdForBigBlind(TableCreatedEvent tableCreatedEvent,
            HandDealtEvent handDealtEvent) {
        var seatPositionToPlayerIdMap = tableCreatedEvent.getSeatPositionToPlayerMap();
        return seatPositionToPlayerIdMap.get(handDealtEvent.getBigBlindPosition());
    }

}

File: src/test/kotlin/com/flexpoker/table/command/aggregate/testhelpers/TableTestUtils.kt
package com.flexpoker.table.command.aggregate.testhelpers

import com.flexpoker.table.command.Card
import com.flexpoker.table.command.CardRank
import com.flexpoker.table.command.HandRanking
import com.flexpoker.table.command.PocketCards
import com.flexpoker.table.command.aggregate.DefaultRandomNumberGenerator
import com.flexpoker.table.command.aggregate.HandEvaluation
import com.flexpoker.table.command.aggregate.TableState
import com.flexpoker.table.command.aggregate.applyEvents
import com.flexpoker.table.command.aggregate.eventproducers.createTable
import com.flexpoker.table.command.aggregate.eventproducers.startNewHandForNewGame
import com.flexpoker.table.command.commands.CreateTableCommand
import com.flexpoker.table.command.events.HandDealtEvent
import com.flexpoker.table.command.events.TableCreatedEvent
import com.flexpoker.table.command.events.TableEvent
import com.flexpoker.test.util.datageneration.CardGenerator.createPocketCards1
import com.flexpoker.test.util.datageneration.CardGenerator.createPocketCards2
import com.flexpoker.test.util.datageneration.DeckGenerator.createDeck
import java.util.UUID

fun createBasicTable(tableId: UUID, vararg playerIdsArray: UUID): TableState {
    val playerIds = setOf(*playerIdsArray)
    val command = CreateTableCommand(tableId, UUID.randomUUID(), playerIds, 6)
    return applyEvents(createTable(command.tableId, command.gameId, command.numberOfPlayersPerTable,
        command.playerIds, DefaultRandomNumberGenerator()))
}

fun createBasicTableAndStartHand(tableId: UUID?, vararg playerIdsArray: UUID): List<TableEvent> {
    val playerIds = setOf(*playerIdsArray)
    val smallBlind = 10
    val bigBlind = 20
    val shuffledDeckOfCards = ArrayList<Card>()
    val cardsUsedInHand = createDeck()
    val handEvaluation1 = HandEvaluation()
    handEvaluation1.playerId = playerIdsArray[0]
    handEvaluation1.handRanking = HandRanking.FLUSH
    handEvaluation1.primaryCardRank = CardRank.KING
    val handEvaluation2 = HandEvaluation()
    handEvaluation2.playerId = playerIdsArray[1]
    handEvaluation2.handRanking = HandRanking.STRAIGHT
    handEvaluation2.primaryCardRank = CardRank.KING
    val handEvaluations = HashMap<PocketCards, HandEvaluation>()
    handEvaluations[createPocketCards1()] = handEvaluation1
    handEvaluations[createPocketCards2()] = handEvaluation2

    val command = CreateTableCommand(tableId!!, UUID.randomUUID(), playerIds, 6)
    val initEvents = createTable(command.tableId, command.gameId, command.numberOfPlayersPerTable,
        command.playerIds, DefaultRandomNumberGenerator())
    val initState = applyEvents(initEvents)
    val events = startNewHandForNewGame(initState, smallBlind, bigBlind, shuffledDeckOfCards, cardsUsedInHand,
        handEvaluations, DefaultRandomNumberGenerator())

    return initEvents + events
}

fun fetchIdForButton(tableCreatedEvent: TableCreatedEvent, handDealtEvent: HandDealtEvent): UUID {
    val seatPositionToPlayerIdMap = tableCreatedEvent.seatPositionToPlayerMap
    return seatPositionToPlayerIdMap[handDealtEvent.buttonOnPosition]!!
}

fun fetchIdForSmallBlind(tableCreatedEvent: TableCreatedEvent, handDealtEvent: HandDealtEvent): UUID {
    val seatPositionToPlayerIdMap = tableCreatedEvent.seatPositionToPlayerMap
    return seatPositionToPlayerIdMap[handDealtEvent.smallBlindPosition]!!
}

fun fetchIdForBigBlind(tableCreatedEvent: TableCreatedEvent, handDealtEvent: HandDealtEvent): UUID {
    val seatPositionToPlayerIdMap = tableCreatedEvent.seatPositionToPlayerMap
    return seatPositionToPlayerIdMap[handDealtEvent.bigBlindPosition]!!
}

fun blindPlayerIds(state: TableState): Triple<UUID, UUID, UUID> {
    return Triple(
        state.seatMap.entries.first { it.key == state.buttonOnPosition }.value!!,
        state.seatMap.entries.first { it.key == state.smallBlindPosition }.value!!,
        state.seatMap.entries.first { it.key == state.bigBlindPosition }.value!!
    )
}
